HW2/P5/physics.pyx

```diff
@@ -55,46 +64,83 @@ cdef void sub_update(FLOAT[:, ::1] XY,
                      float R,
                      int i, int count,
                      UINT[:, ::1] Grid,
-                     float grid_spacing) nogil:
+                     float grid_spacing,
+                     int grid_size,
+                     omp_lock_t *locks) nogil:
     cdef:
         FLOAT *XY1, *XY2, *V1, *V2
-        int j, dim
+        int j, dim, x, y
         float eps = 1e-5
+        int pos_grid_x, pos_grid_y
 
     # SUBPROBLEM 4: Add locking
+    acquire(&(locks[i]))
     XY1 = &(XY[i, 0])
     V1 = &(V[i, 0])
+    
+    # Find the postion of the current ball i in the Grid, make sure to make unsigned int!
+    pos_grid_x = <int> (XY1[0] / grid_spacing)
+    pos_grid_y = <int> (XY1[1] / grid_spacing)
+
     #############################################################
     # IMPORTANT: do not collide two balls twice.
     ############################################################
+
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+    
+    # Ensure that the grid positions around the ball are not out of the grid
+    for x in range(max(pos_grid_x-3, 0), min(grid_size, pos_grid_x+3+1)):
```

Comment:
This is quite clean syntax to use max(pos_grid..0) to make sure it's within the grid, I ended up using if statements which is more clunky